rust/datafusion/src/physical_plan/distinct_expressions.rs

```diff
@@ -0,0 +1,303 @@
+
+/// Enumeration of types that can be accumulated into a distinct set of values.
+#[derive(Debug, PartialEq, Eq, Hash, Clone)]
+enum DistinctScalarValue {
+    Int8(i8),
+    Int16(i16),
+    Int32(i32),
+    Int64(i64),
+    UInt8(u8),
+    UInt16(u16),
+    UInt32(u32),
+    UInt64(u64),
+}
+
+/// For a given expression, maps its Arrow DataType into a LargeList of the
+/// same DataType.
+fn list_of(expr: Arc<dyn PhysicalExpr>, input_schema: &Schema) -> Result<DataType> {
+    let value_data_type = expr.data_type(input_schema)?;
+
+    match value_data_type {
+        DataType::Int8
+        | DataType::Int16
+        | DataType::Int32
+        | DataType::Int64
+        | DataType::UInt8
+        | DataType::UInt16
+        | DataType::UInt32
+        | DataType::UInt64 => Ok(DataType::LargeList(Box::new(value_data_type))),
+        _ => Err(ExecutionError::NotImplemented(
+            "Unsupported column data type for DISTINCT".to_string(),
+        )),
+    }
+}
+
+fn accumulate_scalar(
+    accum: &mut FnvHashSet<DistinctScalarValue>,
+    value: Option<ScalarValue>,
+) -> Result<()> {
+    let accum_value = match value {
+        Some(ScalarValue::Int8(v)) => Some(DistinctScalarValue::Int8(v)),
+        Some(ScalarValue::Int16(v)) => Some(DistinctScalarValue::Int16(v)),
+        Some(ScalarValue::Int32(v)) => Some(DistinctScalarValue::Int32(v)),
+        Some(ScalarValue::Int64(v)) => Some(DistinctScalarValue::Int64(v)),
+        Some(ScalarValue::UInt8(v)) => Some(DistinctScalarValue::UInt8(v)),
+        Some(ScalarValue::UInt16(v)) => Some(DistinctScalarValue::UInt16(v)),
+        Some(ScalarValue::UInt32(v)) => Some(DistinctScalarValue::UInt32(v)),
+        Some(ScalarValue::UInt64(v)) => Some(DistinctScalarValue::UInt64(v)),
+        Some(ScalarValue::Null) => None,
+        _ => {
+            return Err(ExecutionError::NotImplemented(
+                "Unsupported scalar value for DISTINCT accumulator".to_string(),
+            ))
+        }
+    };
+
+    match accum_value {
+        Some(v) => {
+            accum.insert(v);
+        }
+        None => {}
+    }
+
+    Ok(())
+}
+
+macro_rules! accum_batch {
+    ($ARRAY_TY:ident, $DISTINCT_SCALAR_TY: path, $ARRAY: expr, $ACCUM: expr) => {{
+        let array = $ARRAY.as_any().downcast_ref::<$ARRAY_TY>().ok_or_else(|| {
+            ExecutionError::ExecutionError("Error downcasting array".to_string())
+        })?;
+
+        for i in 0..array.len() {
+            $ACCUM.insert($DISTINCT_SCALAR_TY(array.value(i)));
+        }
+
+        Ok(())
+    }};
+}
+
+#[derive(Debug)]
+struct DistinctValuesAccumulator {
+    values: FnvHashSet<DistinctScalarValue>,
+}
+
+impl Accumulator for DistinctValuesAccumulator {
+    fn accumulate_scalar(&mut self, value: Option<ScalarValue>) -> Result<()> {
+        accumulate_scalar(&mut self.values, value)
+    }
+
+    fn accumulate_batch(&mut self, _array: &ArrayRef) -> Result<()> {
+        Err(ExecutionError::NotImplemented(
```

Comment:
Certainly. This path isn't reachable in the scenarios this PR addresses, but I think this will be relevant to `SELECT COUNT(DISTINCT col)` queries that does not include a `GROUP BY`.